### PR TITLE
First iteration of VQ_VAE

### PR DESCRIPTION
**Pull request by TobiasNorlund**

### CODE REVIEW

File: src/vqvae/vqvae.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+
+        if global_step % 100 == 0:
+            self.logger.experiment.add_image('image', torch.cat([x, x_rec], dim=-1)[0], global_step)
+        global_step += 1
+
+        beta = 0.25
+        loss = rec_loss + emb_loss + beta * com_loss
+        return loss
+
+    def configure_optimizers(self):
```
Comment (Filco306) [inline src/vqvae/vqvae.py:173]: add a way to configure lr here? And perhaps use same as in the paper, `2e-4`?

File: src/vqvae/vqvae.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+        x = self._conv_trans_1(x)
+        x = F.relu(x)
+        
+        return self._conv_trans_2(x)
+
+
+
+class VQVAE(pl.LightningModule):
+    
+    def __init__(self):
```
Comment (Filco306) [inline src/vqvae/vqvae.py:111]: I think we should make it modifiable i.e., allow input arguments such as `num_hiddens` etc.

File: src/vqvae/vqvae.py [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+            nn.Conv2d(in_channels=num_residual_hiddens,
+                      out_channels=num_hiddens,
+                      kernel_size=1, stride=1, bias=False)
+        )
+    
+    def forward(self, x):
+        return x + self._block(x)
+
+
+class ResidualStack(nn.Module):
```
Comment (Filco306) [inline src/vqvae/vqvae.py:28]: Nice one

File: src/vqvae/vqvae.py [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+
+class VQVAE(pl.LightningModule):
+    
+    def __init__(self):
+        super().__init__()
+        self.encoder = Encoder(
+            in_channels=3,
+            num_hiddens=128,
+            num_residual_layers=2,
+            num_residual_hiddens=32,
```
Comment (Filco306) [inline src/vqvae/vqvae.py:117]: In the paper, they say: `The encoder consists of 2 strided convolutional layers with stride 2 and window size 4 × 4, followed by two residual 3 × 3 blocks (implemented as ReLU, 3x3 conv, ReLU, 1x1 conv), all having 256 hidden units`.  Is this correct if you are reimplementing the paper? I might be tired and wrong, but I believe your architecture differs. Not that there is anything wrong with that, just whether we should reimplement the paper or not

File: src/vqvae/vqvae.py [hunk 5]
```diff
@@ -0,0 +1,188 @@
+# %%
+import pytorch_lightning as pl
```
Comment (Filco306) [inline src/vqvae/vqvae.py:2]: Very nice. I did not know about this package

File: src/vqvae/vqvae.py [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+        #optimizer = torch.optim.SGD(self.parameters(), 1e-3)
+        return optimizer
+
+#%%
+
+from pl_bolts.datamodules import CIFAR10DataModule
+datamodule = CIFAR10DataModule('/workspace/data', batch_size=32)
+
+global global_step
+global_step = 0
```
Comment (Filco306) [inline src/vqvae/vqvae.py:184]: Just as a general thing but is having a global var avoidable? perhaps have it as a variable inside the model, `self.global_step`?

File: src/vqvae/vqvae.py [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+
+    def configure_optimizers(self):
+        optimizer = torch.optim.Adam(self.parameters(), lr=1e-3)
+        #optimizer = torch.optim.SGD(self.parameters(), 1e-3)
+        return optimizer
+
+#%%
+
+from pl_bolts.datamodules import CIFAR10DataModule
+datamodule = CIFAR10DataModule('/workspace/data', batch_size=32)
```
Comment (Filco306) [inline src/vqvae/vqvae.py:181]: Again, just a detail but if we are reimplement the paper they use a batch-size of 128, if we want the same `lr` as them

File: src/vqvae/vqvae.py [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+        global global_step
+        x, y = inputs
+        x_rec, z_e, e = self(x)
+
+        rec_loss = F.mse_loss(x, x_rec)
+        emb_loss = F.mse_loss(z_e.detach(), e)
+        com_loss = F.mse_loss(z_e, e.detach())
+
+        self.log("rec_loss", rec_loss)
+        self.log("emb_loss", emb_loss)
```
Comment (Filco306) [inline src/vqvae/vqvae.py:162]: middle term optimized by the embedding only (only writing for my own understanding)

File: src/vqvae/vqvae.py [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+    def training_step(self, inputs, _):
+        global global_step
+        x, y = inputs
+        x_rec, z_e, e = self(x)
+
+        rec_loss = F.mse_loss(x, x_rec)
+        emb_loss = F.mse_loss(z_e.detach(), e)
+        com_loss = F.mse_loss(z_e, e.detach())
+
+        self.log("rec_loss", rec_loss)
```
Comment (Filco306) [inline src/vqvae/vqvae.py:161]: optimized by decoder + enc (only writing for my own understanding)

File: src/vqvae/vqvae.py [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+        self.log("emb_loss", emb_loss)
+        self.log("com_loss", com_loss)
```
Comment (Filco306) [inline src/vqvae/vqvae.py:163]: optimized by dec

Comment (Filco306) [reply src/vqvae/vqvae.py:163]: commitment loss

File: src/vqvae/vqvae.py [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+
+        return x_rec, z_e, e
+
+    def training_step(self, inputs, _):
+        global global_step
+        x, y = inputs
+        x_rec, z_e, e = self(x)
+
+        rec_loss = F.mse_loss(x, x_rec)
+        emb_loss = F.mse_loss(z_e.detach(), e)
```
Comment (Filco306) [inline src/vqvae/vqvae.py:158]: `sg[z_e(x)]`. Nice

File: src/vqvae/vqvae.py [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+            num_hiddens=128,
+            num_residual_layers=2,
+            num_residual_hiddens=32,
+        )
+        self.conv_emb = nn.Conv2d(
+            in_channels=128,
+            out_channels=128,
+            kernel_size=1,
+            stride=1)
+        self.embeddings = nn.Embedding(512, 128)
```
Comment (Filco306) [inline src/vqvae/vqvae.py:124]: From the paper: `In our experiments we define N discrete latents (e.g., we use a field of 32 x 32 latents for ImageNet,
or 8 x 8 x 10 for CIFAR10)`. Do they mean then that their discrete latents have a dimensionality of `640`?

Comment (Filco306) [reply src/vqvae/vqvae.py:124]: Because for ImageNet, they later say that their discrete space is `32 x 32 x 1` with `K=512` i.e., their `D = 32 x 32 x 1` = 1024, right?

File: src/vqvae/vqvae.py [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+        z_e = self.conv_emb(z_e)
+        z_e = z_e.permute((0, 2, 3, 1))  # (b, h, w, c)
+
+        sq_dist = (z_e[:, :, :, None, :] - self.embeddings.weight).pow(2).sum(dim=-1)  # (b, h, w, 512)
+        e_index = sq_dist.argmin(dim=-1)  # (b, h, w) 
+        e = self.embeddings(e_index)  # (b, h, w, c)
+
+        z_q = z_e + (e - z_e).detach()  # Copy gradient from z_q to z_e  (b, h, w, c)
+
+        z_q = z_q.permute((0, 3, 1, 2))  # (b, c, h, w)
```
Comment (Filco306) [inline src/vqvae/vqvae.py:144]: Just so I know what is going on here, why permuting here?

Comment (Filco306) [reply src/vqvae/vqvae.py:144]: And on the 2 lines below

**Review by Filco306 — APPROVED** [review]
Overall I think you have done an awesome work :D I just have some tiny comments, but my major discussion I want to have with you is regarding the architecture choice and whether you are trying to reproduce the paper; because in that case I think your architecture might not be entirely correct? I can be wrong however and I can have missed something most definitely. 

Either way, strong LGTM and good work! :D

File: src/vqvae/vqvae.py (hunk 4 above)
Comment (TobiasNorlund) [reply src/vqvae/vqvae.py:117]: Yeah, we first implemented the architecture from the paper (https://github.com/TobiasNorlund/vq-vae/blob/8883039a78cb798f8b3c292348ad180df28f8384/src/vqvae/vqvae.py#L41) but changed to the architecture used [here](https://github.com/zalandoresearch/pytorch-vq-vae/blob/master/vq-vae.ipynb). I don't think it matters very much tbh

File: src/vqvae/vqvae.py (hunk 6 above)
Comment (TobiasNorlund) [reply src/vqvae/vqvae.py:184]: Yeah this is a hack to get the step count into the train_step, can try to make it cleaner :+1:

File: src/vqvae/vqvae.py (hunk 12 above)
Comment (TobiasNorlund) [reply src/vqvae/vqvae.py:124]: So 8x8x10 actually means they get 10 different codes for each patch, using 10 different codebooks (see [this link](https://github.com/zalandoresearch/pytorch-vq-vae/blob/master/vq-vae.ipynb)) I haven't implemented multiple codebooks yet

File: src/vqvae/vqvae.py (hunk 13 above)
Comment (TobiasNorlund) [reply src/vqvae/vqvae.py:144]: Just to move the channel dimension back into its original place
